### DISCUSSION TITLE
role pxe_stack and role time  don't work anymore with skip tag

### DISCUSSION THREAD

**Discussion by jpm38** [bluebanquise/bluebanquise#508] [General]
We use to run these roles with  the skip option : --skip package,service
The role fails with undefined variable.

Looking into the role we have:
pxe_stack/tasks/main.yml:

name: include_vars ░ Gather OS specific variables
include_vars: "{{ item }}"
with_first_found:

files:
"vars/{{ ansible_facts.distribution | replace(' ','') }}{{ ansible_facts.distribution_major_version }}.yml"
"vars/{{ ansible_facts.os_family }}_{{ ansible_facts.distribution_major_version }}.yml"
"vars/{{ ansible_facts.distribution | replace(' ','_') }}.yml"
"vars/{{ ansible_facts.os_family }}.yml"
skip: true
tags:
firewall
internal
package
service
template

Some tags have been added to the task to allow (I suppose) these variables to be defined when executing pxe_stack role with some tags.

So IMHO, this is a bad practice to add these tags here.

**Comment by btravouillon** [comment]
> Some tags have been added to the task to allow (I suppose) these variables to be defined when executing pxe_stack role with some tags.

That's right, prior to this change the variables were not included when ansible-playbook was called with --tags sometag, which led several tasks depending on those variables to fail.
This change was discussed in code review of #434.

My first proposal was to use the tag `always`, which did not reach consensus.
Then we either had to support --tags xor --skip-tags. We went with --tags. I would be glad to review any better solution if you have an idea in mind.

**Reply by jpm38** [reply]
I've also in mind to use another specific tag (always as instance) for the include_vars task.

**Reply by btravouillon** [reply]
There is the tag `internal` for this task, but `--skip-tags` seems to have precedence over `--tags`.

**Comment by oxedions** [comment]
I will propose a change on this for the next release (1.5).

Point is the following:

The current **Gather OS** came with a nice feature: capability to provide variables at multiple levels (distribution, os_familly, version, etc).

But after month of usages, I see some drawbacks:

1. No way to precedence them (I still do not understand why Ansible set them so high in precedence). This is not in the philosophy of BlueBanquise
2. Tags issues (your issue, and need for support, etc)
3. Execution overhead: when deploying playbooks on hundred of nodes, one task takes some time, so one Gather OS per roles takes a lot of time, for a very simple operation that should not take time

However, I wish to keep the nice feature described above.

Proposal solution would be the following.

Vars are set in defaults/main/*.yml

So for example defaults/main/{RedHat.yml|Ubuntu.yml|...}

In each file, content can be like:

```
myrole_packages_RedHat: # we are using os_familly here
  CentOS:
    _7:
      - c
      - b
    _8:
      - d
      - b
  RedHat:
    - a
    - b
```

Then, content is used this way in package task for example:

```
{{
vars['myrole_packages_'+ansible_facts.distribution][ansible_facts.distribution_major_version] | 
default(vars['myrole_packages_'+ansible_facts.distribution]) |
default(vars['myrole_packages_'+ansible_facts.os_familly][ansible_facts.distribution_major_version]) |
default(vars['myrole_packages_'+ansible_facts.os_familly])
}}
```

If indented properly in task, it is readable like before.

So it loads variables automatically, without tags, overhead, and is "precedencable". It also still brings the nice way to provide multiple levels.

**Reply by jpm38** [reply]
Looks good to me.

**Reply by btravouillon** [reply]
> No way to precedence them (I still do not understand why Ansible set them so high in precedence). This is not in the philosophy of BlueBanquise

This is on purpose and the main difference between `vars` and `defaults` directories. The content of `vars` is for variables internal to the role. This is where the role developer will define the value of some variables used in the tasks.
The content of `defaults` is for end users (sysadmin) that want to change some variables by defining these in the inventory.

I can't see a use case where we should allow end users to override internal variables. The `vars` mostly define the list of packages, services and firewall services. It is possible that some variables defined in `vars` may have to be moved to `defaults`. However, introducing the ability for end users to change those variables by moving all `vars` to `defaults` is a **bad** idea.

> Tags issues (your issue, and need for support, etc)

We definitely should use tag 'always'. The cost is minimal during execution of a playbook.

> Execution overhead: when deploying playbooks on hundred of nodes, one task takes some time, so one Gather OS per roles takes a lot of time, for a very simple operation that should not take time

Did you bench this? What is the cost of the "Gather OS" task at scale?

**Reply by oxedions** [reply]
> I can't see a use case where we should allow end users to override internal variables. The `vars` mostly define the list of packages, services and firewall services. It is possible that some variables defined in `vars` may have to be moved to `defaults`. However, introducing the ability for end users to change those variables by moving all `vars` to `defaults` is a **bad** idea.

Why a so bad idea ? BlueBanquise was founded with in mind the idea of giving full powers to sys admin. End user sys admin should have the full rights to say "I wish to use this package instead of this one, etc" for services that do not require configuration files. For example, in the pxe_stack role, it is possible to choose between atftp or fbtftp service as default tftp server. What is not cool is that even playbook vars cannot precedence vars/include_vars.

Still, I also understand your POV. If we want to prevent end users to override these variables, we can place these files in vars/main/*.yml. They will be auto-loaded, and it also solves remaining issues.

> We definitely should use tag 'always'. The cost is minimal during execution of a playbook.

This will trigger all includes when just executing a single role/tag of a playbook. Impossible 🙁 

> Did you bench this? What is the cost of the "Gather OS" task at scale?

Never recorded figures, but I have seen it live taking time on large clusters (around 300-400 nodes). This was not an issue when deploying first time since most of the time was taken by packages installation, but when just { updating | ensuring configuration is ok for all}, it was not very cool to spend so much time for something that should be instantaneous.

**Reply by oxedions** [reply]
j2 example. Core only, addons are addon dev choice:

```
[root@management1 tasks]# cat /etc/bluebanquise/internal/group_vars/all/j2_variables/roles_shortcuts.yml
---

j2_role_packages: "{{ vars[role_name + '_packages_' + ansible_facts.os_family][ansible_facts.distribution]['_' + ansible_facts.distribution_major_version] | default(vars[role_name + '_packages_' + ansible_facts.os_family][ansible_facts.distribution]) | default(vars[role_name + '_packages_' + ansible_facts.os_family][ansible_facts.os_family]['_' + ansible_facts.distribution_major_version]) | default(vars[role_name + '_packages_' + ansible_facts.os_family][ansible_facts.os_family]) }}"

j2_role_services: "{{ vars[role_name + '_services_' + ansible_facts.os_family][ansible_facts.distribution]['_' + ansible_facts.distribution_major_version] | default(vars[role_name + '_services_' + ansible_facts.os_family][ansible_facts.distribution]) | default(vars[role_name + '_services_' + ansible_facts.os_family][ansible_facts.os_family]['_' + ansible_facts.distribution_major_version]) | default(vars[role_name + '_services_' + ansible_facts.os_family][ansible_facts.os_family]) }}"
[root@management1 tasks]#
```

Can be used as a shortcut to simplify roles syntax, not mandatory.
I am deeply thinking of introducing these two j2 variables (packages and services). This would solve all.

**Reply by johnnykeats** [reply]
Agree with both.

These values should not be in `defaults` since there is a need to explicitely separate what is intended to be updated by users and what should not (wish to change something in `vars`: fork).

But includes vars indeed generate important side effects.

j2_ variables seems the proper way to do it: most core roles already rely on others j2.

**Reply by btravouillon** [reply]
> Why a so bad idea ?

Because any user will be able to replace httpd with nginx, or rsyslog with syslog-ng, etc. The roles were written and tested to support some packages/services. This will be a nightmare for support teams.

**Comment by oxedions** [comment]
Here is the best way I found 😊 

A j2 variable like this one:

```
j2_role_variables: "{{ vars[role_name + '_' + ansible_facts.distribution + '_' + ansible_facts.distribution_major_version] | default(vars[role_name + '_' + ansible_facts.distribution]) | default(vars[role_name + '_' + ansible_facts.os_family + '_' + ansible_facts.distribution_major_version]) | default(vars[role_name + '_' + ansible_facts.os_family]) }}"
```

So this is generic. For pxe_stack role for example, file vars/main/RedHat_8.yml could be:

```
---
pxe_stack_RedHat_8:
    apache_cgi_module: cgid
    apache_conf_path: /etc/httpd/conf.d
    apache_user: apache
    packages_to_install:
      - ipxe-x86_64-bluebanquise
      - ipxe-arm64-bluebanquise
      - httpd
      - python3-policycoreutils
      - clustershell
      - python3-clustershell
      - policycoreutils
    default_tftp_package: atftp
    services_to_start:
      - httpd
    default_tftp_service: atftpd
    firewall_services_to_add:
      - http
      - tftp
```

Or it could be for whole RedHat a file named vars/main/RedHat.yml with same dict but named pxe_stack_RedHat. Etc.

And, for example, packages_to_install is directly reachable using: `"{{ j2_role_variables['packages_to_install '] }}"`

**Reply by btravouillon** [reply]
It seems a little bit over-engineered to me. What are we trying to resolve?

>> We definitely should use tag 'always'. The cost is minimal during execution of a playbook.

> This will trigger all includes when just executing a single role/tag of a playbook. Impossible

That's definitely possible, it will only load variables and won't change anything on the target systems. This is harmless.
What would be the performance cost? Are we talking about 5+ minutes? Is it a common use case to run a playbook with tens of roles while executing a single role with --tags? Or is it a corner case?

If some roles have to be run more than others, I would recommend users to split their playbooks per usage. I admit I am not fan of the one-size-fits-all playbook with abusive use of tags. I do prefer smaller playbooks, per function/role. It's still possible to create a generic playbook that will import other plays.

I do prefer to keep things simple: maintain roles in their respective directories, use `inventory` for user data and `internal` inventory to compute data from user data. Any performance issue that could be solved with some playbook/usage improvements must be solved by playbook/usage improvements.

Once again, what are we trying to resolve?

**Reply by bouriquet** [reply]
Hi,

Agree with the last solution provided by @oxedions 

> That's definitely possible, it will only load variables and won't change anything on the target systems. This is harmless.
> What would be the performance cost? Are we talking about 5+ minutes? Is it a common use case to run a playbook with tens of roles while executing a single role with --tags? Or is it a corner case?

It does affect report readability when using ansible-playbook cmd. Not sure about performance issue, but it's common to run playbooks with tags or skip-tags, and anyway i think we can't get rid of this feature. Typical use case would be the diskless customization skipping "identify" tags and using this same tag to configure compute nodes once the boot.
That's only an example, we are using them in many scenario, as a temporary deployment (i.e. bootstrap) before applying the full playbook.

> If some roles have to be run more than others, I would recommend users to split their playbooks per usage. I admit I am not fan of the one-size-fits-all playbook with abusive use of tags. I do prefer smaller playbooks, per function/role. It's still possible to create a generic playbook that will import other plays.

Not sure i understand your point on this, depending on what you call "function/role". Once your cluster is configured, you should be able to run any playbook and don't ask to yourself or your administrator "should i run this X playbook on theses nodes or the master one ? what have been applied to my cluster ?" it breaks the statefull side of playbooks.

**Reply by btravouillon** [reply]
> Not sure i understand your point on this, depending on what you call "function/role". Once your cluster is configured, you should be able to run any playbook and don't ask to yourself or your administrator "should i run this X playbook on theses nodes or the master one ? what have been applied to my cluster ?"

Example was: _This will trigger all includes when just executing a single role/tag of a playbook._
In such case a smaller playbook might fit the need. It's still possible to write a master playbook that imports smaller playbooks. If the admin wants to apply a subset of roles, they can run the smaller playbook that matches the function/role. I'm almost sure sysadmin know which playbook to run depending on what they want to achieve on their cluster.

If the only issue is report readability, I confirm this solution is over-engineered. Just my thought.

**Reply by oxedions** [reply]
Then maybe best solution for everyone is to revert to a KISS state:

Files in `vars/main/*.yml`
Each file is called `'distribution' + '_' + 'distribution_major_version'.yml`

In each file, content is:

```
---
'role_name' + '_' + 'distribution' + '_' + 'distribution_major_version':
    apache_user: apache
    packages_to_install:
      - ipxe-x86_64-bluebanquise
      - ipxe-arm64-bluebanquise
      ...
```

And usage is:

```
"{{ vars[role_name + '_' + ansible_facts.distribution + '_' + ansible_facts.distribution_major_version]['packages_to_install'] }}"
```

This solution is very simple and easy to debug.

I agree the mechanism was interesting, but maybe all of this is too much for such a simple HPC stack. Lets keep KISS way to ensure maximum compatibility with everyone.

**Reply by johnnykeats** [reply]
Every step that simplify the stack is a good move. Let's do that.

If the stack is on the way to a diet, maybe the #418 could be a good candidate. Is ssh_jump really a need? ssh_master role is known to be sensitive, adding more features not needed inside the template will be a burden.

**Reply by oxedions** [reply]
> If the stack is on the way to a diet, maybe the #418 could be a good candidate. Is ssh_jump really a need? ssh_master role is known to be sensitive, adding more features not needed inside the template will be a burden.

Why do you hate me so much...
So maybe beginning of 1.5 will be the time to clean code and remove non necessary things. Maybe it is good time to time to do some kind of "KISS pass" on the stack 😐
